fix: restart greeter when helper is in a wrong state

ItsShamed opened the pull request:
This PR fixes a logic bug that prevents SDDM from restarting the greeter when either the session exits with code 1, or the SDDM helper gets signal-killed.

### Helper side

If the helper gets killed by a signal (currently handled are `SIGTERM` and `SIGHUP`), exit with `Auth::HELPER_OTHER_ERROR`. This will prevent the helper from exiting with code `Auth::HELPER_AUTH_ERROR`, which is wrong.
If the session exits with a bad code (!= 0), exit with `Auth::HELPER_SESSION_ERROR`. This will prevent the helper from exiting `Auth::HELPER_AUTH_ERROR` if the session deliberately ends with code 1.

### Daemon side

**DO NOT** cast the `exitCode` passed by `QProcess::finished` as a `Auth::HelperStatus` if the `exitStatus` is not `QProcess::NormalExit`. On Linux `QProcess::finished` is like a wrapper around `waitid(2)` which will put the signal number in `exitCode` if it got killed by one.
The [Qt Docs] even advises against doing it:

> exitCode is the exit code of the process **(only valid for normal exits)**

[Qt Docs]: https://doc.qt.io/qt-6/qprocess.html#finished

- Closes #1806 
- Closes #1908
- Closes #2027

For issues like #2003, this will result in a "restart loop" (which is essentially the same original symptom if the greeter fails to start in the first place).

Yes for the other session related issues, these are not SDDM related issues per se, but I think it's better UX to restart the greeter rather than leaving the user on a blank tty if a session fails.

ItsShamed committed:
fix(helper): do not badly exit on SIGHUP

Signed-off-by: tsrk. <tsrk@tsrk.me>

ItsShamed committed:
fix(helper): avoid exiting 1 at all costs when session finishes

Signed-off-by: tsrk. <tsrk@tsrk.me>

ItsShamed committed:
fix(auth): do not cast exit status if `QProcess` is not exited normally

A QProcess::CrashExit happen in most cases when a signal killed it. And
returning from a signal-killed process on Linux makes QProcess:finished
return the signal, which is a really bad idea to cast into a random
enum.

Qt also mentions that `exitCode` is only valid for QProcess::NormalExit
invocations

Signed-off-by: tsrk. <tsrk@tsrk.me>

Vogtinator commented:
There's actually one more case of this: https://github.com/sddm/sddm/blob/08a4a77073ffc5566baa54fa5345f20eea999648/src/helper/UserSession.cpp#L50

That connects `QProcess::finished(int exitCode, QProcess::ExitStatus exitStatus)` to `UserSession::finished(int)`, losing `exitStatus` and making `exitCode` potentially invalid. Could you fix that up as well?

ItsShamed committed:
fix(helper): remove overload of `QProcess::finished` in UserSession

This will allow HelperApp to dinstinguish between a crash (most likely
by a signal) from a normal exit.

Signed-off-by: tsrk. <tsrk@tsrk.me>

ItsShamed commented:
Should be good now

https://github.com/sddm/sddm/blob/3fd079f1e9bec2b124b61ad2a0d3bd82963d3a91/src/helper/UserSession.cpp#L159

For this line, I think this is fine, as we explicitly specify an enum, and `HelperApp` filters exit codes anyway.
If this is still bothering, I can change to a straight `exit(2)` instead

```diff
diff --git a/src/helper/UserSession.cpp b/src/helper/UserSession.cpp
index 5fd378a..eb6b598 100644
--- a/src/helper/UserSession.cpp
+++ b/src/helper/UserSession.cpp
@@ -156,7 +156,7 @@ namespace SDDM {
                 }
             }
         } else {
-            Q_EMIT finished(Auth::HELPER_OTHER_ERROR, QProcess::NormalExit);
+            exit(Auth::HELPER_OTHER_ERROR);
         }
     }
 
```

Vogtinator commented on src/helper/HelperApp.cpp:
```diff
@@ -193,8 +199,17 @@ namespace SDDM {
+            qWarning("Session crashed (killed by signal %d).", exitCode);
+            exit(Auth::HELPER_SESSION_ERROR);
+        }
+        else if (exitCode != 0) {
```
Shouldn't this check for `NormalExit`?

ItsShamed replied:
The first `if` branch already checks for `CrashExit` instead, if we run into this branch then we should be in the case of a `NormalExit` (as there are only two enum entries in `QProcess::ExitStatus`).
I doubt this enum would change in QtBase for future releases.

Vogtinator replied:
It's still weird to read. Especially if it says that `exitCode` is only valid for `NormalExit`, the code should check for `NormalExit` explicitly.

ItsShamed replied:
I'll rearrange a bit to make it clearer then

ItsShamed committed:
style(helper): rearrange if conditions to explicitly check for normal exit

Signed-off-by: tsrk. <tsrk@tsrk.me>

Vogtinator commented on src/helper/HelperApp.cpp:
```diff
@@ -34,6 +34,7 @@
 #include <QtNetwork/QLocalSocket>
 
 #include <iostream>
+#include <qobject.h>
```
Wrong kind of include. What is this for? Maybe `#include <QtCore/QObject>` but that should be there already.

ItsShamed replied:
oops, probably some LSP shenanigans that I forgot to monitor

Vogtinator reviewed (commented):
LGTM otherwise. FWICT this is effectively a noop in the "happy path", so not that likely to cause regressions.

ItsShamed commented:
Actually, hold off merging this one. I might've actually introduced new issues where SDDM would restart for seemingly no reason (opposite problem). I'm poking around to see what the root cause might be.

ItsShamed commented:
Okay, so these are the logs that made me realize there is a problem.
<img width="784" height="99" alt="image" src="https://github.com/user-attachments/assets/1f46ec2b-e5d1-456f-9868-b64cb3b22af6" />

A process cannot "crash" with `si_status = 0`, since UNIX signals start at 1 with `SIGHUP`.

I went back on Qt Base to see the behavior of waiting for a child process and
https://github.com/qt/qtbase/blob/46dcf3abc2ab4c6fbf14b5e108c585ba65060806/src/corelib/io/qprocess_unix.cpp#L1312-L1316
is absolute ~~dogshit~~ nightmare. There are no checks of the return code and `errno` except for their case of waiting for `EINTR` (`do {} while (ret = -1 && errno = EINTR)`). Which means that this is total undefined behavior if `errno` is anything other than `EINTR` (for example `ECHILD`), and can lead to `QProcess::finished` passing around undefined values.

We can however prevent this by disconnecting the `QProcess::finished` signal when we receive the first one, since those undefined behaviors are more likely to occur after the child initially exited.

Vogtinator commented:
> I went back on Qt Base to see the behavior of waiting for a child process and
https://github.com/qt/qtbase/blob/46dcf3abc2ab4c6fbf14b5e108c585ba65060806/src/corelib/io/qprocess_unix.cpp#L1312-L1316
is absolute dogshit nightmare. There are no checks of the return code and errno except for their case of waiting for EINTR (do {} while (ret = -1 && errno = EINTR)). Which means that this is total undefined behavior if errno is anything other than EINTR (for example ECHILD), and can lead to QProcess::finished passing around undefined values.

If you have a concrete case where this can fail, please report this to Qt upstream.

Vogtinator commented on src/helper/HelperApp.cpp:
```diff
@@ -193,8 +198,28 @@ namespace SDDM {
-    void HelperApp::sessionFinished(int status) {
-        exit(status);
+    void HelperApp::sessionFinished(int exitCode, QProcess::ExitStatus exitStatus) {
+        disconnect(m_session, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished), this, &HelperApp::sessionFinished);
```
This can never be reached unless something went really wrong: All paths in this method lead to `exit`, i.e. it will never return.

If `exit` somehow spawns an event loop again (exit handlers?), that's broken and will cause many "fun" issues due to accidental reentrancy.

ItsShamed replied:
That's what I thought too, all paths lead to an `exit`, but…

```diff
diff --git a/src/helper/HelperApp.cpp b/src/helper/HelperApp.cpp
index f7aa306..6860535 100644
--- a/src/helper/HelperApp.cpp
+++ b/src/helper/HelperApp.cpp
@@ -199,7 +199,8 @@ namespace SDDM {
     }
 
     void HelperApp::sessionFinished(int exitCode, QProcess::ExitStatus exitStatus) {
-        disconnect(m_session, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished), this, &HelperApp::sessionFinished);
+        static int calls = 0;
+        qWarning("Call #%d: HelperApp::sessionFinished(%d, %d)", ++calls, exitCode, exitStatus);
         if (exitStatus == QProcess::NormalExit) {
             if (exitCode != 0) {
                 qWarning("Session crashed (exit code %d).", exitCode);
```

<img width="830" height="128" alt="image" src="https://github.com/user-attachments/assets/53628275-52b5-423d-b0d2-1d042d38e4ca" />

… `HelperApp::sessionFinished` is in fact called multiples times here…

And that's most likely what happened [to me, above](https://github.com/sddm/sddm/pull/2103#issuecomment-3136871058) too.

I don't think this is in our control, again, the code in Qt Base is very sketchy, and I don't trust it to do things correctly. So we might as well take precautions to prevent such cases.

I'll try filling a bug report upstream.

ItsShamed replied:
Oh well: https://bugreports.qt.io/browse/QTBUG-119081
At this point, I don't have hopes this will get fixed whatsoever…
The reasoning is kinda stupid imho…

Vogtinator replied:
> The reasoning is kinda stupid imho…

The reasoning is:

> The fcntl can only fail if the file descriptor is invalid. That means QProcess passed a bad file descriptor, so the bug would be in QProcess.

i.e. if this actually happens, it's a bug in `QProcess` and you can report it.

Vogtinator replied:
> So we might as well take precautions to prevent such cases.

IMO the root cause needs to be found and it's not in `QProcessPrivate::waitForDeadChild`.

ItsShamed replied:
> The fcntl can only fail if the file descriptor is invalid. That means QProcess passed a bad file descriptor, so the bug would be in QProcess.

Leaving out the fact that `fcntl` is not the only reason that `forkfd_wait` can return -1: https://github.com/qt/qtbase/blob/46dcf3abc2ab4c6fbf14b5e108c585ba65060806/src/3rdparty/forkfd/forkfd_linux.c#L224-L229, and that maybe an actual `waitid(2)` syscall with an expired PID could happen.

> IMO the root cause needs to be found and it's not in `QProcessPrivate::waitForDeadChild`.

You might be right, but it still feels wrong to not check errors because of some assumptions…

Anyway, I can't reproduce the "crash with 0" bug anymore, but I'm still not confident since I still get double `sessionFinished` calls despite `exit`s…

Vogtinator replied:
> Anyway, I can't reproduce the "crash with 0" bug anymore, but I'm still not confident since I still get double sessionFinished calls despite exits…

You can try to trace why each of those calls happen.

ItsShamed replied:
So I just realized the `exit` we've been calling, is not the `exit(2)` syscall, but `QCoreApplication::exit` which does in fact returns, and I was able to reproduce the "crash with 0" bug. I concluded this had nothing to do with `QProcess`.

Sorry for this nothingburger, here is what I understood what happened:
- When `UserSession` `QProcessed::finished`, we catch it in `HelperApp::sessionFinished`
- If it's a `QProcess::NormalExit`, call **`QCoreApplication::exit`**:
  - This only signals the event loop that we requested an exit, and the return code we exit with
  - It actually returns in order to process more events
- This means we accidentally fall through our second case of `QProcess::CrashExit`, believing we stumbled into one:
  - By calling this `exit`, this sets the event loop return code to 3
- Our `HelperApp` then process all the remaining events
- Finally, the event loop returns to the main thread
- When `main()` reaches out of scope, it calls our destructor `~HelperApp()`
- Which in turns calls `UserSession::stop()`.
- However, the `UserSession` is already `QProcess::NotRunning` by this stage, so it emits another `QProcess::finished(Auth::HELPER_SESSION_ERROR, QProcess::NormalExit)`.
  - `HelperApp::sessionFinshed` gets called again…

<details>
<summary>Sequence diagram</summary>

```mermaid
sequenceDiagram
    participant main as ::main
    participant HelperApp as SDDM::HelperApp
    participant Backend as SDDM::Backend
    participant UserSession as SDDM::UserSession

    main->>HelperApp: exec()
    Note over HelperApp: setUp() and doAuth()
    HelperApp->>Backend: openSession()
    Backend->>UserSession: start()

    Note over UserSession: Session goes on…
    UserSession->>HelperApp: QProcess::finished()
    Note over HelperApp: sessionFinished()
    opt QProcess::NormalExit
        alt exitCode not 0
            HelperApp->>HelperApp: exit(Auth::HELPER_SESSION_ERROR)
        else
            HelperApp->>HelperApp: exit(Auth::HELPER_SUCCES)
        end
    end
    Note right of HelperApp: Those RETURN and fall through (not exit syscall)
    HelperApp->>HelperApp: exit(Auth::HELPER_SESSION_ERROR)
    HelperApp-->HelperApp: Finish processing events
    HelperApp-->>main: Event loop returns with 3
    main->>main: Return 3
    Note over main: At SCOPE FINISH
    main->>HelperApp: ~HelperApp() (call destructor)
    HelperApp->>UserSession: stop()
    Note over UserSession: state() is QProcess::NotRunning
    UserSession->>HelperApp: QProcess::finished(3, 0)
    Note over HelperApp: Repeat sessionFinished() a second time
```
</details>

Should be pretty simple to fix:
- We should either explicitly `return` in `HelperApp::sessionFinished`, or wrap the second case in an `else` branch
- In the destructor, call `m_session->stop()` only if `m_session->state() != QProcess::NotRunning`.

Vogtinator replied:
> So I just realized the exit we've been calling, is not the exit(2) syscall, but QCoreApplication::exit which does in fact return

In that case it should probably be made a bit more explicit by calling `QCoreApplication::exit()` instead...

Vogtinator replied:
> However, the UserSession is already QProcess::NotRunning by this stage, so it emits another QProcess::finished(Auth::HELPER_SESSION_ERROR, QProcess::NormalExit).

That always seemed odd to me, maybe we can just remove it?

ItsShamed replied:
> In that case it should probably be made a bit more explicit by calling `QCoreApplication::exit()` instead...

Done

> That always seemed odd to me, maybe we can just remove it?

Replaced by a `qInfo()` log

ItsShamed committed:
fix(helper): return after exits and check for session state in destructor

This avoids calling exit an unnecessary number of times, which can
potentially falsify the real return code, and cause restrat loops

Signed-off-by: tsrk. <tsrk@tsrk.me>

ItsShamed committed:
fix(helper): do not emit finish when stopping already stopped session

Signed-off-by: tsrk. <tsrk@tsrk.me>

ItsShamed committed:
style(helper): explicit usage of `QCoreApplication::exit`

Signed-off-by: tsrk. <tsrk@tsrk.me>

ItsShamed committed:
fix(helper): disconnect `sessionFinished` when exiting

Calling `UserSession::stop` results in a call to `QProcess::terminate`,
which sends a SIGTERM to the session. `QProcess` will catch that and
emit an excess `QProcess::finished(15, 1)` which we don't want to handle
since we are in the process of exiting.

Signed-off-by: tsrk. <tsrk@tsrk.me>

basil commented:
@Vogtinator Friendly note that I tested this patch on Fedora 43 and can confirm it resolves #1908. Looking forward to seeing this PR merged and released!
